File: hard/first-missing-positive/first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        unordered_map<unsigned int, unsigned int> map;
        for (auto& i : nums) {
            if (i > 0) {
                map[i] = (unsigned int)i+1;
            }
        }
        if (map.find(1) == map.end()) {
            return 1;
        }
        int minimum = INT_MAX;
        for (auto& i : map) {
            if (map.find(i.first-1) == map.end()) {
                int j = i.first;
                while (map.find(map[j]) != map.end()) {
                    j = map[j];
                }
                if (map[j] < minimum){
                    minimum = map[j];
                }
            } 
        }
        return minimum;
    }
};

File: hard/first-missing-positive/first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        hashmap = {}
        minimum = 10e9
        for i in nums:
            if i > 0:
                hashmap[i] = i+1
        if 1 not in hashmap:
            return 1
        for i in hashmap:
            if i-1 not in hashmap:
                j = i
                while hashmap[j] in hashmap:
                    j = hashmap[j]
                if hashmap[j] < minimum:
                    minimum = hashmap[j]
        return minimum
